torch/cuda/amp/grad_scaler.py

```diff
@@ -262,26 +264,39 @@ def _unscale_grads_(
                         # For scaled fp16 values, there's a good chance coalescing will cause overflow,
                         # so we should check the coalesced _values().
                         if param.grad.dtype is torch.float16:
-                            param.grad = param.grad.coalesce()
+                            param.grad = self._sparse_coalesce(param.grad)
                         to_unscale = param.grad._values()
                     else:
                         to_unscale = param.grad
-
                     # TODO: is there a way to split by device and dtype without appending in the inner loop?
                     per_device_and_dtype_grads[to_unscale.device][
                         to_unscale.dtype
                     ].append(to_unscale)
 
             for device, per_dtype_grads in per_device_and_dtype_grads.items():
                 for grads in per_dtype_grads.values():
-                    torch._amp_foreach_non_finite_check_and_unscale_(
```

Comment:
amp/GradScaler assume gpu tensors
ShardedGradScaler add support for cpu tensors because of cpu offloading

Reply:
Ah, thanks for clarifying!